### DISCUSSION TITLE
How to decode serial data from log.

### DISCUSSION THREAD

**Discussion by spudwebb** [zwave-js/zwave-js#7157] [Q&A] [answered]
Is there an easy way to decode the serial data that one can see in the log file at debug level, when the node use S0 or S2?

Context:  I'm investigating an issue with a lock (Kwikset 910) that is sending a door lock logging report with an "empty" status. Since it is marked as empty zwavejs does not parse any other information from the message sent. I suspect this lock to send a valid report in which only the status is wrong, because the same device was working with another driver. So, I would like to confirm this by manually parsing the message to check what data (timestamp, event type, user code) it contains.

```
2024-09-12T14:35:21.667Z SERIAL « 0x012f00a8000001004c2198810452043a03a036e03c87addd9941566e85b4ee9e7 (49 bytes)
                                  cfe98cc8ffbb27a074c3100b2007f7fd8
2024-09-12T14:35:21.668Z SERIAL » [ACK]                                                                   (0x06)
2024-09-12T14:35:21.671Z DRIVER « [Node 076] [REQ] [BridgeApplicationCommand]
                                  │ RSSI: -78 dBm
                                  └─[SecurityCCCommandEncapsulation]
                                    │ sequenced: false
                                    └─[DoorLockLoggingCCRecordReport]
                                        record #: 6 (empty)
```

**Comment by AlCalzone** [comment, marked as answer]
You can set the `NODE_ENV` environment variable to `development`, then additional info including the plaintext is printed. In Z-Wave JS UI, this works by executing the following driver function:
```ts
process.env.NODE_ENV = "development"
```

**Reply by kpine** [reply]
Are there any downsides to setting this in normal usage?

**Reply by spudwebb** [reply]
thanks, but it doesn't work for S0, right?

**Reply by AlCalzone** [reply]
> Are there any downsides to setting this in normal usage?

Even larger logs than normally.

> thanks, but it doesn't work for S0, right?

Oh, you might be right. There's a workaround though:
1. Clone and build this repo:
```
git clone --filter=blob:none https://github.com/zwave-js/node-zwave-js
yarn bootstrap
yarn build
```

2. Open it in VSCode
3. Open `test/decodeMessage.ts`
4. Ignore the S2-related stuff, edit these lines. `data` is the entire serial frame.
![grafik](https://github.com/user-attachments/assets/fc680aaa-48ce-41f6-a83a-6bf14cd487c6)
5. Set a breakpoint in `packages/cc/src/cc/DoorLockLoggingCC.ts`, after the `super` call in `DoorLockLoggingCCRecordReport`.
6. Set VSCode to debug the "Decode message" task:
![grafik](https://github.com/user-attachments/assets/9368e284-c14a-49c2-b862-d3c4427431bd)
7. Run with <kbd>F5</kbd>. When the breakpoint is hit, you should see the decrypted payload of the report.

**Reply by spudwebb** [reply]
> 1. Clone and build this repo:

Can this repo be built on windows? If so how?

**Reply by AlCalzone** [reply]
It should. You need to have a current version of Node.js installed which has [corepack](https://github.com/nodejs/corepack) included, and you need to run `codepack enable` once to enable it (manages npm/yarn/... automatically for you).

**Reply by AlCalzone** [reply]
I double-checked: The LTS version of Node.js you can download from https://nodejs.org/en works and has corepack support.
Start PowerShell as admin, then run corepack enable. Afterwards it automatically downloads yarn for you when you try to run it:
![grafik](https://github.com/user-attachments/assets/e67969d6-59bf-4e2c-8a5f-ef45d4fdd26f)

**Reply by AlCalzone** [reply]
Looks like there was a problem with the build step. The latest commit on `master` has this fixed.

**Reply by spudwebb** [reply]
thanks, I can now build and run "Decode message" in VSCode, but I'm not sure to understand how to get the `nonce `and `networkKey`.  
Is `networkKey `the S0 key ?
How do I get the `nonce`?

**Reply by AlCalzone** [reply]
Network key is the S0 key. Nonce can be found in the log. It should have been sent to the device just before the report you're trying to decode.

**Reply by spudwebb** [reply]
I finally got it to work, but I had to set 16 bit node IDs in the definition of host in decodeMessage.ts

```
		get nodeIdType() {
			return NodeIDType.Long;
		},
```

**Comment by AlCalzone** [comment]
Showing S0 plaintext in logs is now tracked in https://github.com/zwave-js/node-zwave-js/issues/7162
